benchmarks/benchmark_functions/machine_learning.py

```diff
@@ -17,69 +17,142 @@
+		w_classical.assign_sub(0.05 * grad_class)
+
+
+def _machine_learning_torch(quantum_model, data):
+	"""ML example with torch interface."""
+
+	import torch
+	data = [[torch.tensor(x, dtype=torch.double), torch.tensor(y, dtype=torch.double)] for x, y in data]
+
+	def hybrid_model(x, w_quantum, w_classical):
+		transformed_x = torch.matmul(w_classical, x)
+		return quantum_model(transformed_x, w_quantum)
+
+	def average_loss(w_quantum, w_classical):
+		c = torch.tensor(0, dtype=torch.double)
+		for x, y in data:
+			prediction = hybrid_model(x, w_quantum, w_classical)
+			c += (prediction - y) ** 2
+		return c / len(data)
+
+	n_features = len(data[0][0])
+
+	w_quantum = torch.tensor(random(size=(n_features, n_features)), requires_grad=True, dtype=torch.double)
+	w_classical = torch.tensor(random(size=(n_features, n_features)), requires_grad=True, dtype=torch.double)
+
+	for i in range(50):
+		loss = average_loss(w_quantum, w_classical)
+		loss.backward()
+
+		w_quantum.data -= 0.05 * w_quantum.grad
+		w_classical.data -= 0.05 * w_classical.grad
+		w_quantum.grad = None
+		w_classical.grad = None
 
 
 def benchmark_machine_learning(hyperparams={}, num_repeats=1):
 	"""Trains a hybrid quantum-classical machine learning pipeline.
 
+	The data is generated from Gaussian blobs. The model first multiplies the input vectors with
+	a weight matrix, and then feeds it into a quantum model that uses AngleEmbedding for the encoding
+	and BasicEntanglingLayers as the trainable circuit. The number of qubits and layers correspond to the
+	number of features. Training uses gradient descent with 50 steps.
+
 	Args:
 	hyperparams (dict): hyperparameters to configure this benchmark
 
-		* 'n_wires': Number of wires to use. Will be ignored if custom device and template are provided.
-
-		* 'n_layers': Number of layers in the default template. Will be ignored if custom params are provided.
-
-		* 'diff_method': name of differentiation method
-
-		* 'device': device on which the circuit is run, or valid device name
+		* 'n_features': Number of features of each data sample. Defaults to 4.
 
-		* 'interface': name of the interface to use
+		* 'n_samples': Number of data samples to use. Defaults to 20.
 
-		* 'template': Template to use. The template must take the trainable parameters as its only argument.
+		* 'diff_method': name of differentiation method. Defaults to 'best'.
 
-		* 'params': Numpy array of trainable parameters that is fed into the template.
+		* 'device': device on which the circuit is run, or valid device name. Defaults to 'default.qubit.
 
-		* 'measurement': measurement function like `qml.expval(qml.PauliZ(0)))`
+		* 'interface': name of the interface to use. Defaults to 'autograd'.
 
 	num_repeats (int): How often the same circuit is evaluated in a for loop. Default is 1.
 	"""
 
-	device, diff_method, interface, params, template, measurement = _core_defaults(hyperparams)
+	data, device, diff_method, interface = _ml_defaults(hyperparams)
```

Comment:
Should we move setting defaults to a `setup` function, so we don't time that?  It's probably trivial compared to the rest of the evaluation, but it might help focus the benchmark on what we most care about.

Reply:
Good point...the idea was that someone should be able to run the functions without asv though. Can we use `setup` but still maintain compatibility with other ways to run?

Reply:
Even if you look at `timeit` or other benchmarking tools, they often support a "setup" component.  And if the particular tool does not have a "setup" component, we can just run both functions.